### ISSUE TITLE
Assertion failed: ((!isa<FixedVectorType>(VTy) || (Index + NumSubElts) <= (int)cast<FixedVectorType>(VTy)->getNumElements()) && "SK_InsertSubvector index out of range"), function getInsertSubvectorOverhead

### ISSUE THREAD

**Issue by dcci** [llvm/llvm-project#59098]
```console
davide@mac-4236F6 astencoder_repro % /Users/davide/work/llvm/build-release-assertion/bin/clang -cc1 -triple x86_64-apple-macosx13.0.0 -Os ./ASTEncoder-postprocessed.cpp -vectorize-slp -emit-obj 2>&1 | grep "Assertion"
Assertion failed: ((!isa<FixedVectorType>(VTy) || (Index + NumSubElts) <= (int)cast<FixedVectorType>(VTy)->getNumElements()) && "SK_InsertSubvector index out of range"), function getInsertSubvectorOverhead, file BasicTTIImpl.h, line 153.
```

```console
% cat ASTEncoder-postprocessed.cpp
typedef long __attribute__((ext_vector_type(2))) a;
typedef __attribute__((__ext_vector_type__(4))) __fp16 b;
typedef __attribute__((__ext_vector_type__(8))) __fp16 c;
void d(unsigned);
a e;
void f(unsigned &g) {
  unsigned &i = g;
  struct {
    a bits;
  } j{e};
  d(g);
  c h = (c)j.bits;
  b a = h.hi;
  if (a.y || a.z)
    i = 3;
}
```

**Comment by dcci** [comment]
To reproduce:

```
% /Users/davide/work/llvm/build-release-assertion/bin/clang -cc1 -triple x86_64-apple-macosx13.0.0 -Os ./ASTEncoder-postprocessed.cpp -vectorize-slp -emit-obj
```

**Comment by dcci** [comment]
cc: @fredriss / @ahmedbougacha

**Comment by RKSimon** [comment]
Reduced:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@e = dso_local local_unnamed_addr global <2 x i64> zeroinitializer, align 16

define void @PR59098(ptr nocapture noundef nonnull align 4 dereferenceable(4) %g) {
entry:
  %0 = load <8 x half>, ptr @e, align 16
  %1 = load i32, ptr %g, align 4
  tail call void @d(i32 noundef %1) #2
  %2 = extractelement <8 x half> %0, i64 5
  %tobool = fcmp une half %2, 0xH0000
  %3 = extractelement <8 x half> %0, i64 6
  %tobool3 = fcmp une half %3, 0xH0000
  %or.cond = select i1 %tobool, i1 true, i1 %tobool3
  br i1 %or.cond, label %if.then, label %if.end

if.then:
  store i32 3, ptr %g, align 4
  br label %if.end

if.end:
  ret void
}
declare void @d(i32 noundef)
```
opt -slp-vectorizer
```
opt -slp-vectorizer foo.ll -S -o -
opt: /home/simon/LLVM/llvm-project/llvm/include/llvm/CodeGen/BasicTTIImpl.h:149: llvm::InstructionCost llvm::BasicTTIImplBase<T>::getInsertSubvectorOverhead(llvm::VectorType*, int, llvm::FixedVectorType*) [with T = llvm::X86TTIImpl]: Assertion `(!isa<FixedVectorType>(VTy) || (Index + NumSubElts) <= (int)cast<FixedVectorType>(VTy)->getNumElements()) && "SK_InsertSubvector index out of range"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: opt -slp-vectorizer foo.ll -S -o -
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
opt(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamEi+0x31)[0x561ee4be87f1]
opt(_ZN4llvm3sys17RunSignalHandlersEv+0x44)[0x561ee4be6844]
opt(+0x35e9fc5)[0x561ee4be6fc5]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x14420)[0x7f8dce043420]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb)[0x7f8dcdb1000b]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x12b)[0x7f8dcdaef859]
/lib/x86_64-linux-gnu/libc.so.6(+0x22729)[0x7f8dcdaef729]
/lib/x86_64-linux-gnu/libc.so.6(+0x33fd6)[0x7f8dcdb00fd6]
opt(+0x1cc6cd4)[0x561ee32c3cd4]
opt(+0x1cc76a0)[0x561ee32c46a0]
opt(_ZNK4llvm19TargetTransformInfo14getShuffleCostENS0_11ShuffleKindEPNS_10VectorTypeENS_8ArrayRefIiEENS0_14TargetCostKindEiS3_NS4_IPKNS_5ValueEEE+0x55)[0x561ee3b6b405]
opt(+0x3886bf1)[0x561ee4e83bf1]
opt(_ZN4llvm13slpvectorizer7BoUpSLP12getEntryCostEPKNS1_9TreeEntryENS_8ArrayRefIPNS_5ValueEEE+0x3cc0)[0x561ee4e90170]
```

**Comment by RKSimon** [comment]
CC @alexey-bataev

**Comment by alexey-bataev** [comment]
Fixed in ac93b611659d94fd94b0394e78888b1402cd5199
